Improve cipher generation

Jatewo opened the pull request:
This pull request refactors the cipher implementation to introduce a more extensible class hierarchy for ciphers, adds support for monoalphabetic ciphers alongside homophonic ciphers, and updates the tests and constants accordingly. The main changes include making `Cipher` an abstract base class, implementing two concrete cipher types, updating the main script to support both, and enhancing the test suite for better coverage and clarity.

### Cipher Class Refactoring and Extensibility

* Refactored `Cipher` to be an abstract base class using Python's `ABC` module, and moved shared logic and validation there. Added abstract methods `generate_key` and `encipher` to enforce implementation in subclasses. (`src/encipherment/cipher.py`, [[1]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R7-R11) [[2]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L38-R51) [[3]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L82-R78) [[4]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L91-R87) [[5]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R98-R203)
* Added `HomophonicCipher` and `MonoalphabeticCipher` as concrete subclasses, each with their own `generate_key` and `encipher` logic. (`src/encipherment/cipher.py`, [[1]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51R98-R203) [[2]](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L166-R308)

### Monoalphabetic Cipher Support

* Implemented the `MonoalphabeticCipher` class, which generates a random monoalphabetic substitution key and supports encryption and recurrence encoding. (`src/encipherment/cipher.py`, [src/encipherment/cipher.pyL166-R308](diffhunk://#diff-e644090a9fdd02d5d3727b202eb184c619c660a715438742af120bfa03440c51L166-R308))
* Updated `src/main.py` to add a `generate_monoalphabetic_cipher` function and allow generation of both cipher types. (`src/main.py`, [[1]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312R1-R13) [[2]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312L18-R58)

### Constants and Configuration

* Increased `MAX_DIFFICULTY` to 30 and added `NUM_CIPHERS` constant for batch generation. (`src/utils/constants.py`, [src/utils/constants.pyL4-R7](diffhunk://#diff-3eade7403a090ae0dccbfc16d0fe917dbc9c68b84e8914219c86c24d3f0aaad5L4-R7))

### Test Suite Improvements

* Updated tests to use `HomophonicCipher` and `MonoalphabeticCipher` instead of the base class, and refactored them into a test class for better organization and coverage. Improved assertions for homophone counts, difficulty validation, and recurrence encoding. (`test/encipherment/test_cipher.py`, [[1]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3L3-R6) [[2]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3R15-R19) [[3]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3L26-R39) [[4]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3L47-R93) [[5]](diffhunk://#diff-73ddfa1579b46b001abf7ffe4f9e34da9a505f9b82c8b6d497aca5bec5aa83a3L101-R151)

---

These changes make the cipher codebase more modular and extensible, add support for new cipher types, and improve both configuration options and test coverage.

Jatewo committed:
Refactor cipher module with abstract base class

Introduced an abstract Cipher base class using ABC, separating homophonic and monoalphabetic cipher logic into dedicated subclasses. Improved code structure, validation, and extensibility, and added recurrence encoding and JSON/string representations to the base class.

Jatewo committed:
Refactor cipher tests for Homophonic and Monoalphabetic

Split tests into two classes for HomophonicCipher and MonoalphabeticCipher, updating imports and fixtures. Added new tests for MonoalphabeticCipher, improved coverage for input validation, recurrence encoding, and key structure. Updated assertions and error messages for clarity and flexibility.

Jatewo committed:
Update tests for HomophonicCipher and difficulty argument

Tests now use HomophonicCipher instead of Cipher and pass the difficulty argument explicitly to generate_cipher. This aligns tests with recent changes to cipher implementation and function signatures.

Jatewo committed:
Update cipher constants and add NUM_CIPHERS

Increased MAX_DIFFICULTY to 30 and introduced NUM_CIPHERS constant set to 10,000 for improved configurability in cipher generation.

Jatewo committed:
Add monoalphabetic cipher generation and refactor

Introduces a new function to generate monoalphabetic ciphers and refactors the homophonic cipher generation to accept a difficulty parameter. Updates imports and main block to support both cipher types and improve flexibility.

Jatewo committed:
Refactor cipher base class and update usages

Renamed the base class 'Cipher' to 'SubstitutionCipher' in src/encipherment/cipher.py and updated all references in src/utils/files.py and related test files. Improved docstrings and clarified method responsibilities for cipher classes. This refactor enhances clarity and better reflects the class hierarchy for substitution ciphers.

Jatewo committed:
Add monoalphabetic cipher generation and tests

Implemented the generate_monoalphabetic_cipher function in main.py and added corresponding tests in test_main.py to verify its behavior and error handling. Also refactored imports and improved code formatting for clarity.

Jatewo committed:
fix: linting

Fantaskink commented on src/encipherment/cipher.py:
```diff
@@ -35,24 +37,21 @@ class Cipher:
 
 	PLAINTEXT_PATTERN = re.compile(r"^[a-z]+$")
```
Hardcoded plaintext regex separate from alphabet specified in constants.py

Jatewo replied:
Yeah, that's an old leftover. I have replaced it with an all() check ensuring they are in the alphabet specified in constants.py

Jatewo committed:
Refactor plaintext validation and import ALPHABET constant

Replaces regex-based plaintext validation with a check using the ALPHABET constant for improved clarity and maintainability. Removes unused regex import and updates imports to include ALPHABET from utils.constants.

Mafusn reviewed (approved):
LGTM 🦖

Jatewo committed:
Refactor HomophonicCipher state initialization and tests

HomophonicCipher now initializes key, ciphertext, and recurrence_encoding as empty values in __init__, and sets them during generate_key and encipher. Tests updated to explicitly call generate_key and encipher before assertions, ensuring state is properly set for each test.

Mafusn reviewed (approved):
LGTM 🦘

TheMOHD3886 reviewed (approved):
lgtm 👍
